### PR TITLE
Feature/adding response telemetry

### PR DESCRIPTION
**Pull request by Audacia-RhysSmith**
| Action | Done? |
| --- | --- |
| Code compiles and unit tests all pass locally | ✔ |
| Debug/console log code removed | ✔ |
| Commented out code removed | ✔ |
| Unit tests added/updated | ✔ |
| README updated | ❌ |
| Considered: <br /> - Performance <br /> - Security <br /> - Logging | ✔ |
| Licenses of any new/upgraded dependencies checked, with no copyleft dependencies introduced | ✔ |

image

image

### CODE REVIEW

File: Audacia.Log.AspNetCore/Audacia.Log.AspNetCore.csproj [hunk 1]
```diff
@@ -8,7 +8,7 @@
     <PackageRequireLicenseAcceptance>true</PackageRequireLicenseAcceptance>
     <Copyright>Copyright © Audacia $([System.DateTime]::Now.Year)</Copyright>
     <PackageIcon>package-icon.png</PackageIcon>
-    <Version>3.0.0</Version>
+    <Version>3.1.0</Version>
```
Comment (jackpercy-acl) [inline Audacia.Log.AspNetCore/Audacia.Log.AspNetCore.csproj]: As `AddRequestBodyTelemetry` has been renamed to `AddActionRequestBodyTelemetry` should this actually bump to v4.0.0 as this is a breaking change?

File: Audacia.Log.AspNetCore/HttpDependencyBodyCaptureTelemetryInitializer.cs [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+    public HttpDependencyBodyCaptureTelemetryInitializer(IOptions<LogActionFilterConfig> options)
+    {
+        Configure(options?.Value);
+    }
+
+    /// <summary>
+    /// Applies configuration to the log filter if provided.
+    /// </summary>
+    /// <param name="config">global or action config.</param>
+    private void Configure(LogActionFilterConfig? config)
```
Comment (jackpercy-acl) [inline Audacia.Log.AspNetCore/HttpDependencyBodyCaptureTelemetryInitializer.cs]: Appreciate `LogActionFilterConfig` type is probably being re-used as it uses all the same config, but this is not an action filter - could either rename or create a different config type?

File: Audacia.Log.AspNetCore/HttpDependencyBodyCaptureTelemetryInitializer.cs [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+    {
+        if (config == null)
+        {
+            return;
+        }
+
+        DisableBodyContent = config.DisableBodyContent;
```
Comment (jackpercy-acl) [inline Audacia.Log.AspNetCore/HttpDependencyBodyCaptureTelemetryInitializer.cs]: Don't think this is used?

Do we perhaps want config for `DisableResponseBody` and `DisableRequestBody` so you only have to log one or the other? Or two separate initializers, and you add one or both?

File: Audacia.Log.AspNetCore/README.md [hunk 4]
```diff
@@ -96,11 +96,42 @@ public IConfiguration Configuration { get; set; }
 public void ConfigureServices(IServiceCollection services)
 {
   services.ConfigureActionContentLogging(Configuration);
-  services.AddRequestBodyTelemetry();
+  services.AddActionRequestBodyTelemetry(Configuration);
```
Comment (jackpercy-acl) [inline Audacia.Log.AspNetCore/README.md]: This method does not take a param

File: Audacia.Log.AspNetCore/README.md [hunk 5]
```diff
@@ -96,11 +96,42 @@ public IConfiguration Configuration { get; set; }
...
+
+```csharp
+using Audacia.Log.AspNetCore;
+
+public IConfiguration Configuration { get; set; }
+
+public void ConfigureServices(IServiceCollection services)
+{
+  services.ConfigureActionContentLogging(Configuration);
+  services.AddActionResponseBodyTelemetry(Configuration);
```
Comment (jackpercy-acl) [inline Audacia.Log.AspNetCore/README.md]: This method does not take a param

File: Audacia.Log.AspNetCore/README.md [hunk 6]
```diff
@@ -96,11 +96,42 @@ public IConfiguration Configuration { get; set; }
...
+
+```csharp
+using Audacia.Log.AspNetCore;
+
+public IConfiguration Configuration { get; set; }
+
+public void ConfigureServices(IServiceCollection services)
+{
+  services.ConfigureDependencyBodyContentLogging(Configuration);
+  services.AddDependencyBodyTelemetry(Configuration);
```
Comment (jackpercy-acl) [inline Audacia.Log.AspNetCore/README.md]: This method does not take a param

File: Audacia.Log.AspNetCore/ServiceCollectionExtensions.cs [hunk 7]
```diff
@@ -88,18 +113,64 @@ public static IServiceCollection AddRequestBodyTelemetry(this IServiceCollection
         return services;
     }
 
+    /// <summary>
+    /// Configures logging of the http bodies for all http dependency requests.
+    /// </summary>
+    /// <exception cref="ArgumentNullException"><paramref name="services"/> is <see langword="null"/>.</exception>
+    /// <exception cref="ArgumentNullException"><paramref name="configuration"/> is <see langword="null"/>.</exception>
+    public static IServiceCollection ConfigureDependencyBodyContentLogging(
```
Comment (jackpercy-acl) [inline Audacia.Log.AspNetCore/ServiceCollectionExtensions.cs:134]: I think a change to this method make sense with a new config type like a mentioned previously, currently this does the exact same as `ConfigureActionContentLogging`, even using the same config section name.

File: Audacia.Log.AspNetCore/README.md [hunk 8]
```diff
@@ -114,9 +145,9 @@ To configure the overrides for "sub" and "role" add the `LogActionFilter` sectio
 }
 ```
 
-##### Configure global request filtering
+##### Configure global request and response filtering
```
Comment (jackpercy-acl) [inline Audacia.Log.AspNetCore/README.md]: This section should mention it also configures the dependency tracking logs, unless you add a new section/type in which case that should be documented too

File: Audacia.Log.AspNetCore/README.md [hunk 9]
```diff
@@ -96,11 +96,42 @@ public IConfiguration Configuration { get; set; }
...
 {
   services.ConfigureActionContentLogging(Configuration);
-  services.AddRequestBodyTelemetry();
+  services.AddActionRequestBodyTelemetry(Configuration);
   services.AddControllers(x => x.Filters.Add<LogRequestBodyActionFilterAttribute>());
 }
 ```
 
+#### ResponseBodyActionLogFilter
+This filter can be registered to include logs for the beginning and end of each Action Method. This only includes response body as well as details of the response such as the type of model returned. Register it like so:
```
Comment (jackpercy-acl) [inline Audacia.Log.AspNetCore/README.md]: Appreciate this is a copy paste and slight tweak to the description above, but not really sure this describes what either of these filters do very well.

It doesn't add logs for the beginning and end of each action method? Could say "This filter can be registered to include the response body to each action" or something? Can change the the one for request body too?

File: Audacia.Log.AspNetCore/RedactionDictionary.cs [hunk 10]
```diff
@@ -10,41 +10,44 @@
 namespace Audacia.Log.AspNetCore;
 
 /// <summary>
-/// Creates a dictionary of form values on the action, excluding parameters marked as personal information.
+/// Creates a dictionary of key value pairs, excluding parameters marked as personal information.
 /// </summary>
-public sealed class ActionArgumentDictionary : Dictionary<string, object>
+public sealed class RedactionDictionary : Dictionary<string, object>
```
Comment (jackpercy-acl) [inline Audacia.Log.AspNetCore/RedactionDictionary.cs:17]: Nice rename :+1:

**Review by jackpercy-acl — CHANGES_REQUESTED** [review]

File: Audacia.Log.AspNetCore/HttpDependencyBodyCaptureTelemetryInitializer.cs (hunk 2 above)
Comment (Audacia-RhysSmith) [reply Audacia.Log.AspNetCore/HttpDependencyBodyCaptureTelemetryInitializer.cs]: Yeah I think we should probably create a separate config object. Can add.

File: Audacia.Log.AspNetCore/Audacia.Log.AspNetCore.csproj (hunk 1 above)
Comment (Audacia-RhysSmith) [reply Audacia.Log.AspNetCore/Audacia.Log.AspNetCore.csproj]: Yeah that's valid, can up to be on version 4.

Comment (Audacia-RhysSmith) [reply Audacia.Log.AspNetCore/Audacia.Log.AspNetCore.csproj]: Bumped.

File: Audacia.Log.AspNetCore/HttpDependencyBodyCaptureTelemetryInitializer.cs (hunk 2 above)
Comment (Audacia-RhysSmith) [reply Audacia.Log.AspNetCore/HttpDependencyBodyCaptureTelemetryInitializer.cs]: Split out.

File: Audacia.Log.AspNetCore/HttpDependencyBodyCaptureTelemetryInitializer.cs (hunk 3 above)
Comment (Audacia-RhysSmith) [reply Audacia.Log.AspNetCore/HttpDependencyBodyCaptureTelemetryInitializer.cs]: Added more granual control.

File: Audacia.Log.AspNetCore/README.md (hunk 4 above)
Comment (Audacia-RhysSmith) [reply Audacia.Log.AspNetCore/README.md]: Thanks for spotting, removed.

File: Audacia.Log.AspNetCore/README.md (hunk 9 above)
Comment (Audacia-RhysSmith) [reply Audacia.Log.AspNetCore/README.md]: Yeah good point I'll modify it and we can review to see if it is more fitting?

File: Audacia.Log.AspNetCore/README.md (hunk 5 above)
Comment (Audacia-RhysSmith) [reply Audacia.Log.AspNetCore/README.md]: Removed - thanks for spotting.

File: Audacia.Log.AspNetCore/README.md (hunk 6 above)
Comment (Audacia-RhysSmith) [reply Audacia.Log.AspNetCore/README.md]: Removed - thanks for spotting.

File: Audacia.Log.AspNetCore/README.md (hunk 8 above)
Comment (Audacia-RhysSmith) [reply Audacia.Log.AspNetCore/README.md]: Have split it out the config, and added a new section for configuring the dependencies. Can this stay the same?

File: Audacia.Log.AspNetCore/ServiceCollectionExtensions.cs (hunk 7 above)
Comment (Audacia-RhysSmith) [reply Audacia.Log.AspNetCore/ServiceCollectionExtensions.cs:134]: Updated to use different config location.

File: Audacia.Log.AspNetCore.Tests/Audacia.Log.AspNetCore.Tests.csproj [hunk 11]
```diff
@@ -5,16 +5,19 @@
...
+    <EnforceCodeStyleInBuild>true</EnforceCodeStyleInBuild>
+    <RunAnalyzersDuringBuild>true</RunAnalyzersDuringBuild>
+    <GenerateDocumentationFile>true</GenerateDocumentationFile>
     <IsPackable>false</IsPackable>
     <IsTestProject>true</IsTestProject>
   </PropertyGroup>
 
   <ItemGroup>
-    <PackageReference Include="Audacia.CodeAnalysis" Version="1.9.0">
+    <PackageReference Include="Audacia.CodeAnalysis" Version="1.11.0">
```
Comment (jackpercy-acl) [inline Audacia.Log.AspNetCore.Tests/Audacia.Log.AspNetCore.Tests.csproj:16]: Thanks for sorting all the code analysis stuff :+1:

File: Audacia.Log.AspNetCore/HttpDependencyBodyCaptureTelemetryInitializer.cs [hunk 12]
```diff
@@ -66,21 +77,21 @@ public HttpDependencyBodyCaptureTelemetryInitializer(IOptions<LogActionFilterCon
...
-    private void Configure(LogActionFilterConfig? config)
+    private void Configure(LogDependencyFilterConfig? config)
     {
         if (config == null)
         {
             return;
         }
 
-        DisableBodyContent = config.DisableBodyContent;
+        MapHttpDependencyConfiguration(config);
```
Comment (jackpercy-acl) [inline Audacia.Log.AspNetCore/HttpDependencyBodyCaptureTelemetryInitializer.cs:87]: :+1: nice

File: Audacia.Log.AspNetCore/HttpDependencyBodyCaptureTelemetryInitializer.cs [hunk 13]
```diff
@@ -32,13 +33,23 @@ public class HttpDependencyBodyCaptureTelemetryInitializer : ITelemetryInitializ
     public int MaxDepth { get; set; } = 32;
 
     /// <summary>
-    /// Gets or sets a value indicating whether the logging of all data in the request body is disabled.
+    /// Gets or sets a value indicating whether the HTTP dependencies should be captured and logged.
     /// </summary>
-    public bool DisableBodyContent { get; set; }
+    public bool DisabledHttpTracking { get; set; }
```
Comment (jackpercy-acl) [inline Audacia.Log.AspNetCore/HttpDependencyBodyCaptureTelemetryInitializer.cs]: Don't think this is actually used? If it is needed, should be renamed `Disable` not `Disabled`. If it is not needed then can remove from `LogDepedencyFilterConfig` too?

File: Audacia.Log.AspNetCore/ServiceCollectionExtensions.cs [hunk 14]
```diff
@@ -136,16 +149,17 @@ public static IServiceCollection ConfigureDependencyBodyContentLogging(
         services.AddSingleton<IHttpContextAccessor, HttpContextAccessor>();
 
         // Configure global configuration for the LogActionFilter
-        services.Configure<LogActionFilterConfig>(configuration.GetSection("LogActionFilter"));
+        services.Configure<LogActionFilterConfig>(configuration.GetSection("LogDependencyFilter"));
```
Comment (jackpercy-acl) [inline Audacia.Log.AspNetCore/ServiceCollectionExtensions.cs]: The generic should be `LogDependencyFilterConfig`

File: Audacia.Log/Audacia.Log.csproj [hunk 15]
```diff
@@ -3,26 +3,28 @@
...
     <Authors>Audacia</Authors>
-    <Description>Standardized SeriLog logging configuration for Audacia projects.</Description>
-    <Copyright>Copyright © Audacia 2020</Copyright>
+    <Description>Standardized logging configuration for Audacia projects.</Description>
+    <Copyright>Copyright © Audacia $([System.DateTime]::UtcNow.Year)</Copyright>
     <PackageLicenseExpression>MIT</PackageLicenseExpression>
     <PackageRequireLicenseAcceptance>true</PackageRequireLicenseAcceptance>
     <PackageIcon>package-icon.png</PackageIcon>
-    <Version>1.1.0</Version>
+    <Version>1.1.1</Version>
```
Comment (jackpercy-acl) [inline Audacia.Log/Audacia.Log.csproj]: This can stay at 1.1.0, the pipeline will bump it

**Review by jackpercy-acl — CHANGES_REQUESTED** [review]

File: Audacia.Log.AspNetCore/HttpDependencyBodyCaptureTelemetryInitializer.cs (hunk 13 above)
Comment (Audacia-RhysSmith) [reply Audacia.Log.AspNetCore/HttpDependencyBodyCaptureTelemetryInitializer.cs]: Updated.

File: Audacia.Log.AspNetCore/ServiceCollectionExtensions.cs (hunk 14 above)
Comment (Audacia-RhysSmith) [reply Audacia.Log.AspNetCore/ServiceCollectionExtensions.cs]: Okay - updated.

File: Audacia.Log/Audacia.Log.csproj (hunk 15 above)
Comment (Audacia-RhysSmith) [reply Audacia.Log/Audacia.Log.csproj]: Reverted.

File: Audacia.Log.AspNetCore/README.md (hunk 9 above)
Comment (jackpercy-acl) [reply Audacia.Log.AspNetCore/README.md]: All the docs look great, thanks Rhys

File: Audacia.Log.AspNetCore/HttpDependencyBodyCaptureTelemetryInitializer.cs (hunk 13 above)
Comment (jackpercy-acl) [reply Audacia.Log.AspNetCore/HttpDependencyBodyCaptureTelemetryInitializer.cs]: @Audacia-RhysSmith don't think anything has changed RE this?

**Review by jackpercy-acl — APPROVED** [review]

**Pull request merged**